migration-0.55/main.go

```diff
@@ -0,0 +1,366 @@
+	resources = removeUnderscoreFields(resources, statement)
+
+	attributes = append(attributes, resources...)
+
+	return attributes, nil
+}
+
+func getTableStatement(conn clickhouse.Conn) (string, error) {
+	ctx := context.Background()
+	statements := []ShowCreateTableStatement{}
+	query := "SHOW CREATE TABLE signoz_logs.logs"
+	err := conn.Select(ctx, &statements, query)
+	if err != nil {
+		zap.S().Error(fmt.Errorf("error while getting logs table statement. Err=%v", err))
+		return "", err
+	}
+	return statements[0].Statement, nil
+}
+
+func hasMaterializedColumn(tableStatement, field, dataType string) bool {
+	// check the type as well
+	regex := fmt.Sprintf("`%s` (?i)(%s) DEFAULT", field, dataType)
+
+	res, err := regexp.MatchString(regex, tableStatement)
+	if err != nil {
+		zap.S().Error(fmt.Errorf("error while matching regex. Err=%v", err))
+		return false
+	}
+	return res
+}
+
+func getClickhouseLogsColumnDataType(dataType string) string {
+	dataType = strings.ToLower(dataType)
+	if dataType == "int64" || dataType == "float64" {
+		return "number"
+	}
+	if dataType == "bool" {
+		return "bool"
+	}
+	return "string"
+}
+
+func addMaterializedColumnsAndAddIndex(conn clickhouse.Conn, fields []LogField) error {
+	ctx := context.Background()
+	for _, field := range fields {
+		// columns name is <type>_<name>_<datatype>
+		colname := fmt.Sprintf("%s_%s_%s", strings.ToLower(field.Type), getClickhouseLogsColumnDataType(field.DataType), strings.ReplaceAll(field.Name, ".", "$$"))
+		keyColName := fmt.Sprintf("%s_%s", field.Type+"s", getClickhouseLogsColumnDataType(field.DataType))
+
+		// create column in logs table
+		for _, table := range []string{"logs_v2", "distributed_logs_v2"} {
+			zap.S().Info(fmt.Sprintf("creating materialized for: %s i.e %s", field.Name, colname))
+			query := fmt.Sprintf("ALTER TABLE signoz_logs.%s on cluster cluster ADD COLUMN IF NOT EXISTS %s %s DEFAULT %s['%s'] CODEC(ZSTD(1))",
+				table, colname, field.DataType, keyColName, field.Name)
+			err := conn.Exec(context.Background(), query)
+			if err != nil {
+				zap.S().Error(fmt.Errorf("error while creating materialized column on logs table. Err=%v", err))
+			}
+
+			query = fmt.Sprintf("ALTER TABLE signoz_logs.%s ON CLUSTER cluster ADD COLUMN IF NOT EXISTS %s_exists bool DEFAULT if(mapContains(%s, '%s') != 0, true, false) CODEC(ZSTD(1))",
+				table,
+				colname,
+				keyColName,
+				field.Name,
+			)
+			err = conn.Exec(ctx, query)
+			if err != nil {
+				zap.S().Error(fmt.Errorf("error while creating exists column on logs table. Err=%v", err))
+				return err
+			}
+		}
+
+		// index not required for bool attributes as the cardinality is only two
+		if strings.ToLower(field.DataType) == "bool" {
+			continue
+		}
+
+		zap.S().Info(fmt.Sprintf("Create index: %s_idx", colname))
+		query := fmt.Sprintf("ALTER TABLE signoz_logs.logs_v2 on cluster cluster ADD INDEX IF NOT EXISTS %s_idx (%s) TYPE bloom_filter(0.01) GRANULARITY 64", colname, colname)
```

Comment:
Is this the index we use in query-service also for number and string types?

Reply:
Yes this is the default we use.